omniscidb/Shared/quantile.h

```diff
@@ -0,0 +1,343 @@
+          } else {
+            ++chunk_idx_;
+            chunk_offs_ = 0;
+            i -= rem;
+          }
+        }
+      }
+      return *this;
+    }
+
+    Iterator operator+(int64_t i) const {
+      Iterator res = *this;
+      res += i;
+      return res;
+    }
+
+    Iterator& operator-=(int64_t i) {
+      if (i < 0) {
+        operator+=(-i);
+      } else {
+        while ((int64_t)chunk_offs_ < i) {
+          --chunk_idx_;
+          i -= chunk_offs_;
+          chunk_offs_ = (*chunks_)[chunk_idx_].size;
+        }
+        chunk_offs_ -= i;
+      }
+      return *this;
+    }
+
+    Iterator operator-(int64_t i) const {
+      Iterator res = *this;
+      res -= i;
+      return res;
+    }
+
+    Iterator& operator++() {
+      ++chunk_offs_;
+      if (chunk_offs_ == (*chunks_)[chunk_idx_].size) {
+        ++chunk_idx_;
+        chunk_offs_ = 0;
+      }
+      return *this;
+    }
+
+    Iterator operator++(int) {
+      Iterator res = *this;
+      operator++();
+      return res;
+    }
+
+    Iterator& operator--() {
+      if (!chunk_offs_) {
+        --chunk_idx_;
+        chunk_offs_ = (*chunks_)[chunk_idx_].size;
+      }
+      --chunk_offs_;
+      return *this;
+    }
+
+    Iterator operator--(int) {
+      Iterator res = *this;
+      operator--();
+      return res;
+    }
+
+    bool operator==(const Iterator& other) const {
+      return chunk_idx_ == other.chunk_idx_ && chunk_offs_ == other.chunk_offs_;
+    }
+
+    bool operator!=(const Iterator& other) const { return !(*this == other); }
+
+    bool operator<(const Iterator& other) const {
+      if (chunk_idx_ != other.chunk_idx_) {
+        return chunk_idx_ < other.chunk_idx_;
+      }
+      return chunk_offs_ < other.chunk_offs_;
+    }
+
+    bool operator<=(const Iterator& other) const {
+      if (chunk_idx_ != other.chunk_idx_) {
+        return chunk_idx_ < other.chunk_idx_;
+      }
+      return chunk_offs_ <= other.chunk_offs_;
+    }
+
+    T& operator*() const {
+      auto data = reinterpret_cast<T*>((*chunks_)[chunk_idx_].data);
+      return data[chunk_offs_];
+    }
+
+   private:
+    const std::vector<Chunk>* chunks_;
+    // Current chunk index. Can be equal to size of chunks_ vector for `end` iterator.
+    size_t chunk_idx_;
+    // Offset in the current chunk. Should always be less than chunk size when the
+    // index points to a valid chunk.
+    size_t chunk_offs_;
+  };
+
+  ChunkedArray(SimpleAllocator* allocator) : allocator_(allocator), cur_idx_(0) {}
+
+  template <typename T>
+  void push(T value) {
+    // Check if we need to allocate a new chunk.
+    if (chunks_.empty() || cur_idx_ == chunks_.back().size) {
+      // Allocator is most probably a RowSetMemoryOwner object. It is not supposed to be
+      // used to allocate very small objects, so we start with 1 KB and double it each
+      // time with 64KB limit.
+      size_t size_to_allocate = std::max((size_t)64, (size_t)1 << chunks_.size()) << 10;
```

Comment:
Missing a shift for `64`?

Reply:
No, looks correct to me. 64 is max number of KB to allocate. The result of `std::max` is shifted by 10 to get size in bytes.

Reply:
Ah right, too many brackets